plugin/pkg/scheduler/algorithm/priorities/priorities.go

```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
 	}
 	return float64(requested) / float64(capacity)
 }
+
+type NodePreferAvoid struct {
+	controllerLister algorithm.ControllerLister
+}
+
+func NewNodePreferAvoidPriority(controllerLister algorithm.ControllerLister) algorithm.PriorityFunction {
+	nodePreferAvoid := &NodePreferAvoid{
+		controllerLister: controllerLister,
+	}
+	return nodePreferAvoid.CalculateNodePreferAvoidPriority
+}
+
+func (npa *NodePreferAvoid) CalculateNodePreferAvoidPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodeLister algorithm.NodeLister) (schedulerapi.HostPriorityList, error) {
+	var score int
+	nodes, err := nodeLister.List()
+	if err != nil {
+		return nil, err
+	}
+
+	result := []schedulerapi.HostPriority{}
+
+	rcs, err := npa.controllerLister.GetPodControllers(pod)
+	if err != nil {
+		for _, node := range nodes.Items {
+			result = append(result, schedulerapi.HostPriority{Host: node.Name, Score: 10})
+		}
+		return result, nil
+	}
+
+	avoidNodes := map[string]bool{}
+	availableNodesNum := len(nodes.Items)
+	for _, node := range nodes.Items {
+		avoidNodes[node.Name] = false
+
+		avoids, err := api.GetAvoidPodsFromNodeAnnotations(node.Annotations)
+		if err != nil {
+			continue
+		}
+		for _, avoid := range avoids.PreferAvoidPods {
+			for _, rc := range rcs {
+				if avoid.PodSignature.PodController.Name == rc.Name && avoid.PodSignature.PodController.Namespace == rc.Namespace {
```

Comment:
You should also check kind. Since you know rc is a ReplicationController, you can just check that avoid.PodSignature.PodController.Kind == "ReplicationController". (Of course when you also add ReplicaSet you would do the same for the ReplicaSets.) I'm actually not sure what to do about version; I'll get back to you about that one later.